### ISSUE TITLE
Single node vs Cluster poor reading performance

### ISSUE THREAD

**Issue by GITHUBKIMIA** [phpredis/phpredis#1259]
Hi there,
We are considering migrating from a classic master - slaves approach to Redis Cluster so we are running performance tests against two different setups: 1- Single Redis node 2- Basic Redis Cluster setup (3 masters/ 3 slaves)

The fact is that reading from the cluster (slaves) takes double time than reading from the single node, whether connections are persistent or not. (Writing also takes longer but our concern is about reading times...)  

I tried several tests, I am pretty sure that the phpredis client is correctly handling the slot map as its requests hit the right slave node (no MOVED answers).

When I run a test querying the single node and a slave node just for the keys contained in this slave then the reading time is almost identical... (Using redis-cli)

Is this the normal behavior, or am I missing something important?   Please find the installation details and the script I am using to check the R/W performance.

Thanks in advance.

- OS: Debian 8 'Jessie'
- Redis: 4.0.1
- PHP: 5.6.30
- phpredis: 3.1.3

```
<?php

//apd_set_pprof_trace();

class RedisTest {

    // redis --> 6379
    // redisCluster --> 7000,7001,7002,7003,7004,7005

    // const MAX_KEYS = 100000;
     const MAX_KEYS = 100000;

    private $_redisCluster;
    private $_redisClusterNopersistent;
    private $_redisStandalone;
    private $_redisStandaloneNopersistent;

    public function __construct()
    {
        $this->_redisStandalone = new Redis();

        $res = $this->_redisStandalone->pconnect('127.0.0.1');

        if (!$res) {
            echo "REDIS CONNECTION new error connect con respuesta: [$res]";
        }

        $this->_redisStandaloneNopersistent = new Redis();

        $res = $this->_redisStandaloneNopersistent->connect('127.0.0.1');

        if (!$res) {
            echo "REDIS CONNECTION new error connect con respuesta: [$res]";
        }

        
        $this->_redisCluster = new RedisCluster(null, array(
            'localhost:7000', 'localhost:7001', 'localhost:7002'), 1.5, 1.5, true);

        $this->_redisCluster->setOption(RedisCluster::OPT_SLAVE_FAILOVER, RedisCluster::FAILOVER_DISTRIBUTE_SLAVES);


        $this->_redisClusterNopersistent = new RedisCluster(null, array(
            'localhost:7000', 'localhost:7001', 'localhost:7002'), 1.5, 1.5, false);

        $this->_redisClusterNopersistent->setOption(RedisCluster::OPT_SLAVE_FAILOVER, RedisCluster::FAILOVER_DISTRIBUTE_SLAVES);
    }

    public function testPerformance()
    {
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisStandalone->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis singleton write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisStandalone->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis singleton read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;

        //*********************************************
        $start = microtime(true);


        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisStandaloneNopersistent->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis singleton no persistent write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisStandaloneNopersistent->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis singleton no persistent read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;

        //*********************************************
        $start = microtime(true);

        //var_dump($this->_redisCluster->_masters());

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisClusterNopersistent->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster no persistent write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisClusterNopersistent->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster no persistent read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;

        //*********************************************
        $start = microtime(true);

        //var_dump($this->_redisCluster->_masters());

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisCluster->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisCluster->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
    }

    public function testFailOver($instance = 0)
    {
        $this->executeOSCommand("sudo service redis_service_700$instance stop");
        $this->executeOSCommand("ps aux | grep redis | grep 700$instance");
        sleep(30);

        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisCluster->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisCluster->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;

        $this->executeOSCommand("sudo service redis_service_700$instance start");
        $this->executeOSCommand("ps aux | grep redis | grep 700$instance");
        sleep(30);
    }

    public function testFailOverNonPersistent($instance = 0)
    {
        $this->executeOSCommand("sudo service redis_service_700$instance stop");
        $this->executeOSCommand("ps aux | grep redis | grep 700$instance");
        sleep(30);

        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisClusterNopersistent->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisClusterNopersistent->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis cluster read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;

        $this->executeOSCommand("sudo service redis_service_700$instance start");
        $this->executeOSCommand("ps aux | grep redis | grep 700$instance");
        sleep(30);
    }


    public function testStandaloneNopersistent()
    {
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $this->_redisStandaloneNopersistent->set('key' . $i, $i);
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis standalone write time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
        $start = microtime(true);

        for ($i = 0; $i < self::MAX_KEYS; $i++){
            $value = $this->_redisStandaloneNopersistent->get('key' . $i);
            if ($value === false) {
                echo "Error accesing key $i" . PHP_EOL;
            }
        }

        $time_elapsed_secs = microtime(true) - $start;
        echo "Redis standalone read time " . $time_elapsed_secs  .  " secs" . PHP_EOL;
    }

    public function executeOSCommand($command, $returnOutput = false)
    {
        try {
            exec($command, $output, $status);
            $stringOutput = implode(';', $output);
            echo "Command: [$command] Output: [$stringOutput] Status: [$status]" . PHP_EOL;
            //if ($status != 0) {
            //    throw new Exception("Error while exec command: [$command] Output:[$stringOutput] Status: [$status]");
            //}
            if ($returnOutput) {
                return $output;
            } else {
                return true;
            }
        } catch (Exception $e) {
            echo "Error while exec command: [$command]: " . $e->getMessage() . PHP_EOL;
        }
    }
}

$test = new RedisTest();

$test->testPerformance();
echo "##############################" .PHP_EOL;
//$test->testPerformance();
echo "##############################" .PHP_EOL;
//$test->testFailOver(0);
//$test->testFailOverNonPersistent(0);
//$test->testStandaloneNopersistent();
//echo "##############################" .PHP_EOL;
//$test->testPerformance();
echo "##############################" .PHP_EOL;
```

**Comment by michael-grunder** [comment]
Hey,

I ran your benchmark script and saw a similar performance dropoff using Redis Cluster.  That said, it gets a bit more interesting.

First, I wrote a small benchmark program in hiredis to eliminate phpredis from the equation entirely, which can be found [here](https://gist.github.com/michael-grunder/5171adc6f28520632fce4166d9d23446).  Following are the averages for different operations using ten runs (raw output [here](https://gist.github.com/michael-grunder/2c74c96e40123caf09f060ea4760b431)).

Type | Op | Keys | Time | Keys/sec
---|---|---|---:|---:
standalone | get | 100000 | 1.5602 | 64172.9
standalone | set | 100000 | 1.77188 | 56898.1
cluster | get | 100000 | 1.80494 | 56357.6
cluster | set | 100000 | 4.04106 | 24807.6

Note that I had to "lock" the keys to a single node by using a `{hash}:N` as hiredis doesn't know anything about redis cluster.

Next I created a program using phpredis similar to the hiredis version with keys distributing across the cluster naturally.   As with my hiredis version, I ran each test ten times. ([code](https://gist.github.com/michael-grunder/b1a13b536f02f46cf74bcc9a75e17bee), [raw output](https://gist.github.com/michael-grunder/1b920d81cfde7dc6fdc8f9159320a15e), averages below):

Type | Op | Keys | Time | Keys/sec
---|---|---:|---:|---:
standalone | get | 100000 | 1.4922 | 67,089
standalone | set | 100000 | 1.5221 | 65,884
cluster | get | 100000 | 2.2985 | 43,545
cluster | set | 100000 | 5.0198 | 19,959

Interestingly there seems to be a really large write performance dropoff in Redis Cluster regardless of which library is being used (I didn't expect it would be that big), and both hiredis and phpredis show a drop in read performance but phpredis' is worse.

That being said, the hiredis program is doing less work.  It's not hashing the key or selecting where to execute commands -- it's simply operating against a single node.

The advantage of Redis cluster however, is that it is distributed so the next thing I tried was running my php test program in parallel ([raw output redis](https://gist.github.com/michael-grunder/d878062e9801c0130d4bbae52e2a6f45), [raw output cluster](https://gist.github.com/michael-grunder/2aefbb8e72ed471cc183f890f0be2434)):

Type | Op | Keys | Time | Keys/sec
---|---|---:|---:|---:
standalone | get | 100000 | 8.7495 | 11,429
standalone | set | 100000| 9.1986 | 10,871
cluster | get | 100000 | 4.8185 | 20,768
cluster | set | 100000 | 7.3722 | 13,565

I ran the php program in parallel like so:
```bash
parallel php-5630 tp2.php redis set 100000 ::: {1..15}
parallel php-5630 tp2.php redis get 100000 ::: {1..15}
parallel php-5630 tp2.php rediscluster set 100000 ::: {1..15}
parallel php-5630 tp2.php rediscluster get 100000 ::: {1..15}
```

Generally speaking I would argue that you will get better "single lane" throughput with standalone redis but higher overall distributed throughput with redis cluster (without having to do any work of sharding the data yourselves).  I was pretty surprised by _how_ much slower write performance is in cluster, but as you mentioned that's not a concern on your end.

I will still do some profiling of redis cluster this weekend to see if there are any obvious places where we're being slow but hopefully this information is at least useful to you :smile:

Cheers,
Mike

**Comment by kjoe** [comment]
I've made some experiment with write performance dropoffs (single client), and I've seen, that on Redis Cluster (and standalone master-slave deployments too), if you add more and more slaves, the write performance continuously slows down (AOF + RDB enabled), especially if you have more then one slave on a host. If you simply stop those slave services, the write performance is increasing to normal level.

**Comment by michael-grunder** [comment]
Hey everyone,

We've added a couple of things to phpredis that improves cluster performance but several times.

TL;DR: [cluster slot caching](https://github.com/phpredis/phpredis/blob/develop/cluster.markdown#slot-caching), combined with persistent connections.

Link to a very quick [benchmark](https://github.com/phpredis/phpredis/issues/825#issuecomment-498054909)
